sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala

```diff
@@ -735,44 +735,67 @@ case class GpuGenerateExec(
       // Project input columns, setting other columns ahead of generator's input columns.
       // With the projected batches and an offset, generators can extract input columns or
       // other required columns separately.
-      val projectedInput = GpuProjectExec.project(input, othersProjectList ++ genProjectList)
-      withResource(projectedInput) { projIn =>
-        // 1. compute split indices of input batch
-        val splitIndices = generator.inputSplitIndices(
-          projIn,
-          othersProjectList.length,
-          outer,
-          new RapidsConf(conf).gpuTargetBatchSizeBytes)
-        // 2. split up input batch with indices
-        makeSplitIterator(projIn, splitIndices).map { splitIn =>
-          withResource(splitIn) { splitIn =>
-            // 3. apply generation on each (sub)batch
-            val ret = generator.generate(splitIn, othersProjectList.length, outer)
-            numOutputBatches += 1
-            numOutputRows += ret.numRows()
-            ret
+      val projectedInput = GpuProjectExec.projectAndCloseWithRetrySingleBatch(
+        SpillableColumnarBatch(input, SpillPriorities.ACTIVE_ON_DECK_PRIORITY),
+        othersProjectList ++ genProjectList)
+      val ret = withResource(SpillableColumnarBatch(projectedInput,
+        SpillPriorities.ACTIVE_ON_DECK_PRIORITY)) { scb =>
+        val targetSizeWrapper = AutoCloseableTargetSize(
+          new RapidsConf(conf).gpuTargetBatchSizeBytes, 16L * 1024 * 1024)
+        withRetry(targetSizeWrapper, splitTargetSizeInHalf) { attempt =>
+          splitAndGenerate(scb, othersProjectList, attempt.targetSize)
+        }.next()
+      }
+      ret.foreach { cb =>
+        numOutputBatches += 1
+        numOutputRows += cb.numRows()
+      }
+      ret.toIterator
+    }
+  }
+
+  // Split up the input batch and call generate on each split.
+  private def splitAndGenerate(scb: SpillableColumnarBatch,
+      othersProjectList: Seq[GpuExpression],
+      targetSize: Long): Seq[ColumnarBatch] = {
+    val splits = withResource(scb.getColumnarBatch()) { cb =>
+      // 1. compute split indices of input batch
+      val splitIndices = generator.inputSplitIndices(
+      cb, othersProjectList.length, outer, targetSize)
+      // 2. split up input batch with indices
+      makeSplits(cb, splitIndices)
+    }
+    withResource(splits) { _ =>
+      splits.safeMap { spillable =>
+        // 3. apply generation on each (sub)batch
+        withRetryNoSplit(spillable) { attempt =>
+          withResource(attempt.getColumnarBatch()) { batch =>
+            generator.generate(batch, othersProjectList.length, outer)
```

Comment:
I _think_ this is the first place with a nested `withRetry`, and I believe that is ok, but would love to see a test in GenerateSuite (since we already have some infra for the scala test) to throw retry and split and retry.

So the way this is going to work if I understand correctly, is that this inner loop under the `withRetryNoSplit` wil go until the split exception, and then the `withRetry` on line 746 should catch it and call back into `splitAndGenerate` with a different target. Does it make sense to combine doGenerateAndClose and splitAndGenerate?

Reply:
I think there is at least one other case of nested retry.  The join code calls `gatherNext` within a `withRetry` block, and that ends up calling `LazySpillableGatherMapImpl.getBuffer`, which has a `withRetyNoSplit` block.

Your understanding is correct.  I could combine `doGenerateAndClose` and `splitAndGenerate`, which is how it was before, but I actually split these out to make the code a little easier to read.

Reply:
Fair enough, and agreed. One deals with changing the target, the other one deals within the target.

Reply:
I will work on adding a test to GenerateSuite as well.